cylc/flow/config_schema.py

```diff
@@ -0,0 +1,612 @@
+        },
+        'events': {
+            'handlers': [VDR.V_STRING_LIST, None],
+            'handler events': [VDR.V_STRING_LIST, None],
+            'startup handler': [VDR.V_STRING_LIST, None],
+            'timeout handler': [VDR.V_STRING_LIST, None],
+            'inactivity handler': [VDR.V_STRING_LIST, None],
+            'shutdown handler': [VDR.V_STRING_LIST, None],
+            'aborted handler': [VDR.V_STRING_LIST, None],
+            'stalled handler': [VDR.V_STRING_LIST, None],
+            'timeout': [VDR.V_INTERVAL],
+            'inactivity': [VDR.V_INTERVAL],
+            'abort if any task fails': [VDR.V_BOOLEAN, False],
+            'abort if startup handler fails': [VDR.V_BOOLEAN],
+            'abort if shutdown handler fails': [VDR.V_BOOLEAN],
+            'abort if timeout handler fails': [VDR.V_BOOLEAN],
+            'abort if inactivity handler fails': [VDR.V_BOOLEAN],
+            'abort if stalled handler fails': [VDR.V_BOOLEAN],
+            'abort on stalled': [VDR.V_BOOLEAN],
+            'abort on timeout': [VDR.V_BOOLEAN],
+            'abort on inactivity': [VDR.V_BOOLEAN],
+        },
+        'reference test': {
+            'expected task failures': [VDR.V_STRING_LIST],
+        },
+        'authorization': {
+            # Allow owners to grant public shutdown rights at the most, not
+            # full control.
+            'public': (
+                [VDR.V_STRING, '']
+                + [level.name.lower().replace('_', '-') for level in [
+                    Priv.IDENTITY, Priv.DESCRIPTION, Priv.STATE_TOTALS,
+                    Priv.READ, Priv.SHUTDOWN]])
+        },
+    },
+    'job platforms': {
+        'default platform': {
+            'run directory': [VDR.V_STRING, '$HOME/cylc-run'],
+            'work directory': [VDR.V_STRING, '$HOME/cylc-run'],
+            'task communication method': [
+                VDR.V_STRING, 'zmq', 'ssh+zmq', 'poll'],
+            'submission polling intervals': [VDR.V_INTERVAL_LIST],
+            'execution polling intervals': [VDR.V_INTERVAL_LIST],
+            'scp command': [
+                VDR.V_STRING, 'scp -oBatchMode=yes -oConnectTimeout=10'],
+            'ssh command': [
+                VDR.V_STRING, 'ssh -oBatchMode=yes -oConnectTimeout=10'],
+            'use login shell': [VDR.V_BOOLEAN, True],
+            'login hosts': [VDR.V_STRING, None],
+            'batch system': [VDR.V_STRING, None],
+            'cylc executable': [VDR.V_STRING, 'cylc'],
+            'global init-script': [VDR.V_STRING],
+            'retrieve job logs': [VDR.V_BOOLEAN],
+            'retrieve job logs command': [VDR.V_STRING, 'rsync -a'],
+            'retrieve job logs max size': [VDR.V_STRING],
+            'retrieve job logs retry delays': [VDR.V_INTERVAL_LIST],
+            'task event handler retry delays': [VDR.V_INTERVAL_LIST],
+            'tail command template': [
+                VDR.V_STRING, 'tail -n +1 -F %(filename)s'],
+            'batch systems': {
```

Comment:
Do we want 1 batch system per job platform so that if the same platform is used with a different system that is a separate platform, or do we want `__MANY__` here. @oliver-sanders @hjoliver?

Reply:
A platform uses a single batch system.

Reply:
Hmm, I guess it's a fairly safe bet that no one would install both PBS and Slurm (say) on the same platform, even though you could do that. It wouldn't make much sense from a resource management perspective!

Reply:
There may be cases where one might want to submit background jobs to a login node (e.g. housekeeping) rather than submitting properly to a compute node. Not the best practice perhaps but a likely workaround given the limitations of some setups.

Reply:
Yeah I do see a lot of that here at NIWA (I briefly forgot that we consider "background" is a batch system :grimacing: )

Reply:
Matt said
> You may have missed my other comments. Each platform should only have one batch system. If there are two batch systems, then you actually have two different platforms (that may happen to share some resources).

So there seem to be two visions 
- Different batch system == different platform, even on the same cluster/machine.
- A platform can have multiple batch systems.

I can't see any major difference in how they are wired, so I want to know is why @matthewrmshin thinks that the first one is better. Are there top level platform settings
that might not be suitable for multiple batch systems? If so, should these be moved into
the `[[batch system info]]` subsection?

Reply:
Summary of in-person discussion on the above:
Each Platform is to be a unique combination of host and batch system, so that if users want to run background jobs on a resource (which in many cases they ought not) that constitutes a different platform to using Slurm or PBS on the same resource. For example you might have

```cfg
[job platforms]
    [[orig]]
        ...
    [[HPC background]]
        ...
    [[HPC PBS]]
        ...
    [[Nutmeg Slurm]]
        ...
    [[Tims Raspberry Pi]]
        ...
```